src/routing/bootstrap.rs

```diff
@@ -0,0 +1,653 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::Comm;
+use crate::{
+    consensus::Proven,
+    crypto,
+    error::{Error, Result},
+    location::DstLocation,
+    messages::{BootstrapResponse, JoinRequest, Message, Variant, VerifyStatus},
+    node::Node,
+    peer::Peer,
+    relocation::{RelocatePayload, SignedRelocateDetails},
+    section::{EldersInfo, Section},
+    SectionProofChain,
+};
+use bytes::Bytes;
+use futures::future;
+use std::{mem, net::SocketAddr};
+use tokio::{
+    sync::{mpsc, oneshot},
+    task,
+};
+use xor_name::Prefix;
+
+/// Bootstrap into the network as an infant node.
+/// When this completes, the returned `IncomingConnections` should be passed to the main `Executor`
+/// instead of creating a new one with `Comm::listen` which would cause temporary disconnect.
+pub(crate) async fn infant(
+    node: Node,
+    comm: &Comm,
+    bootstrap_addr: SocketAddr,
+) -> Result<(Node, Section, qp2p::IncomingConnections)> {
+    // NOTE: when we are bootstrapping a new infant node, there is no `Executor` running yet.
+    // So we create a simple throwaway executor here. It works a bit differently than the main
+    // `Executor`. First, it runs inside a `LocalSet`. This allows us to terminate the whole
+    // executor simply by dropping the `LocalSet` - this simplifies things a bit. Second, we don't
+    // `spawn` separate tasks for each message, but send them to a channel instead. This means that
+    // the whole bootstrapping process runs on a single thread. That should be fine as its quite
+    // sequential in nature anyway (send request, receive response, send request, receive response,
+    // ...).
+
+    let (send_tx, send_rx) = mpsc::channel(1);
+    let (recv_tx, recv_rx) = mpsc::channel(1);
+
+    let state = State::new(node, send_tx, recv_rx)?;
+    let local_set = task::LocalSet::new();
+
+    let (abort_tx, abort_rx) = oneshot::channel();
+    let incoming_connections = comm.listen()?;
+    let receive_handle =
+        local_set.spawn_local(receive_messages(incoming_connections, recv_tx, abort_rx));
+
+    let (node, section) = future::join(
+        send_messages(comm, send_rx),
+        local_set.run_until(state.run(vec![bootstrap_addr], None)),
+    )
+    .await
+    .1?;
+
+    // Abort the `receive_messages` task and extract the `incoming_connections` out of it by
+    // awaiting its join handle. This cannot fail, unless there is a bug in the code, so it's OK
```

Comment:
It would make no difference. This task is only to listen to incoming messages and send them to a channel. It doesn't care about how the messages are handled. 

That said, we currently don't reject join requests. What could happen is that the join request gets lost, or fails to accumulate or similar such things. In that case the bootstrapping process would hang. We used to have a timeout for this, but we decided that it would be better if such timeout was handled by the upper layer. They can do something like this, for example:

```rust
tokio::time::timeout(Duration::from_secs(60), Routing::new(...)).await
```

Reply:
maybe worth to put the above as a comment into the routing/mod.rs , above the Routing struct declaration or `fn new` definition ?

Reply:
Already did.